File: starter-packs/python/network/network.py
import websocket
import json
import ssl
import threading
import time
from typing import List, Optional

from src.bot import MyBot
from core.message import MessageType
from core.action import Action
from network.decoder import JDISDecoder


class Socket:  
    def __init__(self, url: str, token: str):
        self.url = url
        self.token = token
        self.bot = MyBot()
        self.ping_interval = 1

        
    def run(self):
        print(f"Starting bot with base URL: {self.url}, token: {self.token}")
        try:
            ws = websocket.WebSocketApp(self.url,
                                        header={'Authorization': self.token},
                                        on_open=self.on_open,
                                        on_message=self.on_message,
                                        on_error=self.on_error,
                                        on_close=self.on_close)
        except Exception as e:
            print("Error: ", e)
            return
        
        ws.run_forever(sslopt={"cert_reqs": ssl.CERT_NONE})


    def handle_message(self, message: bytes) -> Optional[List[Action]]:
        message_type = int(message[0])
        response = None

        decoder = JDISDecoder()

        if message_type == MessageType.GameStart.value:
            map_state = decoder.decode_map_state(message[1:])
            self.bot.on_start(map_state)

        elif message_type == MessageType.GameState.value:
            game_state = decoder.decode_game_state(message[1:])
            response = self.bot.on_tick(game_state)

        elif message_type == MessageType.GameState.GameEnd.value:
            self.bot.on_end()

        else:
            print("Unknown message type")

        return response


    def on_open(self, ws: websocket.WebSocketApp) -> None:
        print("Connection opened")
        self.start_ping_thread(ws)
        

    def on_message(self, ws: websocket.WebSocketApp, message: bytes) -> None:
        response = self.handle_message(message)
        if response:
            self.send_message(ws, response)
        

    def on_error(self, ws: websocket.WebSocketApp, error: str) -> None:
        print("Error: ", error)
        

    def on_close(self, ws: websocket.WebSocketApp, close_status_code, close_msg) -> None:
        print("Connection closed")
        

    def send_message(self, ws: websocket.WebSocketApp, actions: List[Action]) -> None:
        json_reponse = {}
        for action in actions:
            try:
                json_reponse.update(action.serialize())
            except Exception as e:
                print(e)
    
        json_message = json.dumps(json_reponse)

        print(f"Sending message: {json_message}")
        prefixed_message = bytearray([3]) + json_message.encode('utf-8')
        ws.send(prefixed_message)


    def start_ping_thread(self, ws: websocket.WebSocketApp) -> None:
        ping_thread = threading.Thread(target=self.ping, args=(ws,))
        ping_thread.daemon = True
        ping_thread.start()


    def ping(self, ws: websocket.WebSocketApp) -> None:
        while ws.keep_running:
            ws.send('ping')
            time.sleep(self.ping_interval)
    

File: starter-packs/javascript/network/network.js
import { WebSocket } from 'ws';
import { MyBot } from '../src/bot.js';
import { JDISDecoder } from './decoder.js';
import { MessageType } from '../core/message.js';

class Socket {
  #url;
  #secret;
  #ws;
  #ping_interval = null;
  #bot = null;

  constructor(url, secret) {
    this.#url = url;
    this.#secret = secret;
    this.#bot = new MyBot();
  }


  run() {
    this.#connect();
  }


  #connect() {
    this.#ws = new WebSocket(this.#url, {
      headers: { 'Authorization': this.#secret },
      rejectUnauthorized: false
    });
    this.#ws.binaryType = 'arraybuffer';

    this.#ws.on('open', () => this.#on_open());  
    this.#ws.on('message', (message) => this.#on_message(message));
    this.#ws.on('close', () => this.#on_close());
    this.#ws.on('error', (error) => this.#on_error(error)); 
  }


  #on_open() {
    console.log(`Connected to ${this.#url}`);
    this.#start_heartbeat();
  }

  #on_message(message) {
    const message_type = int(message[0]);
    const decoder = new JDISDecoder();
    
    switch (message_type) {
      case MessageType.GameStart:
        this.#bot.on_start(decoder.decode_map_state(message));
        break;
      case MessageType.GameState:
        const response = this.#bot.on_tick(decoder.decode_game_state(message));
        const actions = new TextEncoder().encode(encode_actions(response));

        const message = new Uint8Array(1 + actions.length);
        message.set(new Uint8Array([3]), 0);
        message.set(actions, 1);
        this.#ws.send(message.buffer);
        break;
      case MessageType.GameEnd:
        this.#bot.on_end();
        break;
    }
  }


  #on_error(error) {
    console.log(`Websocket error: ${error}`);
    this.#stop_heartbeat();
  }


  #on_close() {
    console.log('Websocket connection closed');
    this.#stop_heartbeat();
  }


  #start_heartbeat() {
    this.#ping_interval = setInterval(() => {
      if (this.#ws.readyState === WebSocket.OPEN)
        this.#ws.ping();
    }, 1000);
  }

    
  #stop_heartbeat() {
    if (this.#ping_interval) {
      clearInterval(this.#ping_interval);
      this.#ping_interval = null;
    }
  }
};

/**
 * 
 * @param {Model.Actions} actions 
 */
const encode_actions = (actions) => {
    const data = {};

    actions && actions.forEach(action => {
        if (action === null)
            return;
        
        switch (action.type) {
            case 'dest':
                data[action.type] = action.destination;
    
                break;
            case 'shoot':
                data[action.type] = action.pos;
                break;
            case 'save':
                data[action.type] = btoa(String.fromCharCode.apply(null, action.data));
                break;
            case 'switch':
                data[action.type] = action.weapon;
                break;
            case 'rotate_blade':
                data[action.type] = action.rad;
            default:
                break;
        }
    });

    return JSON.stringify(data);
};


export { Socket };
